### ISSUE TITLE
Making use of functional hooks 

### ISSUE THREAD

**Issue by pfistfl** [mlr-org/mlr3tuning#190]
Creating this issue, so we can discuss whether this makes sense.

pytorch, keras etc. make ample use of [functional hooks/callbacks](https://www.kaggle.com/sironghuang/understanding-pytorch-hooks).

The same principle is used in Keras as  [Callbacks](https://keras.io/callbacks/).
In `keras` you can for example define a function that is run after / before every batch / epoch.

I think this would nicely fit into `mlr3tuning`.

Possible use-cases: 
- After every iteration evaluate the loss on the outer test set (for visualization). 
- Logging evaluated points to a database
- Evaluating predict-step hyperpars i.e. `nrounds* < nrounds` model was trained with.
- Modify a `hyperparam` of the tuner when a certain criterion is met (for `infillCritCB`): 
  - Increase `cb.lambda` after $n$ iterations.
  - Decrease `cb.lambda` when validation error stagnates
- Weave in random evaluations
  

This could also perform as a `Terminator`, as such a hook could basically just set some termination flag.

**Comment by berndbischl** [comment]
in generel yes, we can do this. would it be enough to have this after batch_eval? and i would like to keep this separate from terminators

**Comment by pfistfl** [comment]
Really depends on which use-cases we want to cover.
I guess we should collect a number of situations where we would want this

@mb706

**Comment by berndbischl** [comment]
> Really depends on which use-cases we want to cover.

please outline 1-3 most important ones here. you kinda startet in the OP already with that

**Comment by be-marc** [comment]
@pfistfl https://github.com/mlr-org/bbotk/pull/110 This might cover some of your use-cases. Suggestions are welcome.

**Comment by pfistfl** [comment]
I answered in the PR.

**Comment by be-marc** [comment]
Callbacks are implemented in bbotk and mlr3tuning now.
